source/common/upstream/cluster_manager_impl.cc

```diff
@@ -515,6 +520,31 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
   return true;
 }
 
+// Note: simply deleting the entry corresponding to this cluster from updates_map_
+// isn't enough; the unique_ptrs won't go away because the armed timers hold
+// references to them. So we take care of those too.
+void ClusterManagerImpl::destroyUpdates(const std::string& name) {
```

Comment:
Unless I'm missing something I don't think it needs to be this complicated. Can't you just add a destructor to `PendingUpdates` and call `disableTimer()` from there for the EXPECT_CALL? The timer doesn't hold strong reference to anything.

Reply:
The timer's callback holds a ref to `unique_ptr<PendingUpdates>`... so `~PendingUpdates()` won't be called unless we get rid of the callback, no?

Reply:
@mattklein123 see that we pass `&updates` here:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L422

and the timer is also saved in `updates`....

Reply:
I don't see where this is the case?

Reply:
You are passing around references, not a strong pointer. There is no circular reference that I can see.

Reply:
@mattklein123 hmm, that's not true... that ref won't prevent `~PendingUpdates()` from being called. Sorry, got confused — this can be simplified.